SopraGameAction.cpp

```diff
@@ -0,0 +1,197 @@
+        auto crossedCells = gameController::getAllCrossedCells(this->actor->position, target);
+        std::vector<gameModel::Position> ret;
+        for(const auto &cell : crossedCells){
+            for(const auto &player : env.getOpponents(*actor)){
+                if(player->position == cell){
+                    ret.emplace_back(cell);
+                }
+            }
+        }
+
+        return ret;
+    }
+
+    auto Shot::getAllLandingCells(const gameModel::Environment &env) const -> std::vector<gameModel::Position> {
+#warning Was tun bei n gerade? Dann lässt sich das Quadrat nicht mittig um target platzieren
+        int n = static_cast<int>(std::ceil(gameController::getDistance(actor->position, target) / 7));
+        std::vector<gameModel::Position> ret;
+        using Env = gameModel::Environment;
+        using Cell = gameModel::Cell;
+        auto players = env.getAllPlayers();
+        ret.reserve(n * n);
+        for(int x = target.x - n / 2; x < target.x + n / 2; x++){
+            for(int y = target.y - n / 2; y < target.y + n / 2; y++){
+                if(gameModel::Position{x, y} == target){
+                    continue;
+                }
+
+                if(Env::getCell(x, y) == Cell::OutOfBounds){
+                    continue;
+                }
+
+                if(env.cellIsFree({x, y})){
+                    ret.emplace_back(x, y);
+                }
+            }
+        }
+
+        return ret;
+    }
+
+
+    // fertig
+    auto Move::check(const gameModel::Environment &envi) const -> ActionResult{
+
+        if (gameModel::Environment::getCell(this->target) == gameModel::Cell::OutOfBounds ||
+            gameController::getDistance(this->actor.get()->position, this->target) > 1) {
+            return ActionResult::Impossible;
+        }
+
+        if (this->checkForFoul(envi) == gameModel::Foul::None) {
+            return ActionResult::Success;
+        }
+        else {
+            return ActionResult::Foul;
+        }
+    }
+
+    auto Move::executeAll(const gameModel::Environment &envi) const ->
+        std::vector<std::pair<gameModel::Environment, double>>{
+
+        std::vector<std::pair<gameModel::Environment, double>> resultVect;
+
+        // @toDo: da fehlt noch alles
+
+        return resultVect;
+    }
+
+    Move::Move(std::shared_ptr<gameModel::Player> actor, gameModel::Position target): Action(actor, target) {}
+
+    void Move::execute(gameModel::Environment &envi) const {
+        ActionResult actionResult = this->check(envi);
+
+        switch (actionResult) {
+
+            case ActionResult::Foul :
+                // @ToDo was passiert bei einem Faul???
+
+            case ActionResult::Success :
+
+                if (envi.getPlayer(this->target) != nullptr) {
+                    movePlayerOnEmptyCell(envi, this->target);
+                }
+                this->actor.get()->position = this->target;
```

Comment:
```suggestion
                this->actor->position = this->target;
```
reicht auch da für shared_ptr -> und  * überladen sind